feat: Add Foundry toolkit that enables Sapphire precompile and decryption (#484)
oasisprotocol/sapphire-paratime: integrations/foundry/README.md

# Foundry Package for Sapphire developers

> This project implements Sapphire-compatible precompiles for use with forge tests. It includes support for confidential computing capabilities through encryption/decryption handling.

Reviewer: Wrap to 80 characters.

Author: Wrapped READMEs to 80 characters.

## Overview

This project contains a set of contracts that mock Sapphire's precompile behavior using rust bindings. 
In addition, it contains SapphireDecryptor, a contract that can decrypt encrypted calldata. 
SapphireDecryptor works by inheritence and will forward the decrypted calldata to the contract that called it. 

> This is useful for testing gasless transactions which contain encrypted calldata.

Reviewer: ```suggestion
This is useful, for example, to test the execution of the on-chain generated gasless transactions
with encrypted calldata.
```

Author: Updated the README with suggested changes

> The project provides several key precompiles:

Author: Updated README with suggested changes.

- `SignPrecompile`: Sign messages
- `VerifyPrecompile`: Verify signatures

### Consensus Operations
- `SubcallPrecompile`: Enhanced version with CBOR parsing and state management for:
  - Delegations
  - Undelegations
  - Receipt tracking

## Key Features

1. **Sapphire precompiles as contracts**
   - Can run as native precompiles
   - Easy import into forge tests

2. **Decryption Base contract**
   - Enables decryption at contract level
   - Used as a base contract for other contracts that implement encryption

## Installation
1. **Install dependencies**
    - `forge install foundry-rs/forge-std --no-commit --no-git`
    - `forge install oasisprotocol/sapphire-paratime --no-commit --no-git`
    - Make sure you have rust nightly installed. Go to src/precompiles and build rust bindings:
      - `cargo +nightly build --release` (ignore warnings, this will be fixed in the future)

2. **Run tests**
    
    Project already contains tests for all precompiles (see TestSapphireContracts.t.sol)
    
    Steps to run:
    - Make sure to inherit from SapphireTest, which handles the precompile deployment
    - Run forge tests `forge test`

## Usage

> For a more concise example with fuzz testing, see sapphire-paratime/examples/foundry.

Reviewer: ```suggestion
For a more concise example with fuzz testing, see [sapphire-paratime/examples/foundry].

[sapphire-paratime/examples/foundry]: https://github.com/oasisprotocol/sapphire-paratime/tree/main/examples/foundry
```

Author: Updated README.

# Foundry Package for Sapphire developers

This project implements Sapphire-compatible precompiles 
for use with forge tests. It includes support for 
confidential computing capabilities through encryption/decryption handling.

## Overview

This project contains a set of contracts that mock Sapphire's precompile 
behavior using rust bindings. In addition, it contains SapphireDecryptor, 
a contract that can decrypt encrypted calldata. SapphireDecryptor works by 
inheritence and will forward the decrypted calldata to the contract that 
called it. This is useful, for example, to test the execution of the on-chain 
generated gasless transactions with encrypted calldata.

## List of precompiles

> ### Cryptographic Operations

Reviewer: The line after the title should always be empty.

```suggestion
### Cryptographic Operations

```

- `RandomBytes`: Generate random bytes
- `X25519Derive`: Derive shared secrets using X25519
- `DeoxysiiSeal`: Encrypt data using Deoxys-II
- `DeoxysiiOpen`: Decrypt data using Deoxys-II
- `Curve25519ComputePublic`: Compute public keys
- `DECODE`: decode cbor encoded data and decrypt the calldata. 

> <span style="color:red">Warning</span>: This precompile is not part of 

Author: Updated.

the Sapphire EVM. It is only used for
testing encryption envelope without having to decode the CBOR encoded data 
in solidity tests.

### Key Management
- `KeypairGenerate`: Generate cryptographic keypairs
- `Sign`: Sign messages
- `Verify`: Verify signatures

### Consensus Operations
- `Subcall`: Enhanced version with CBOR parsing and state management for:

>   - Delegations
>   - Undelegations
>   - Receipt tracking

Reviewer: ```suggestion
- Delegations
- Undelegations
- Receipt tracking
```

Author: Removed indent.
